goodtables/compat.py

```diff
@@ -30,22 +31,22 @@
     basestring = basestring
     numeric_types = (int, long, float)
 
-    def csv_reader(data, dialect=csv.excel, **kwargs):
+    def csv_reader(data, dialect=csv.excel, header_index=0, **kwargs):
         """Read text stream (unicode on Py2.7) as CSV."""
 
-        first_lines = list(islice(data, 5))
-        try:
-            dialect = csv.Sniffer().sniff(''.join(first_lines))
-            dialect.delimiter = dialect.delimiter.encode('utf-8')
-            dialect.quotechar = dialect.quotechar.encode('utf-8')
-        except csv.Error:
-            dialect = csv.excel
-
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+        
+        while not re.findall('[^\w ]', dialect.delimiter): 
+            first_lines.extend(list(data.readline()))
+            dialect = detect_dialect(first_lines)
+            
         def iterenc_utf8(data):
             for line in data:
                 yield line.encode('utf-8')
 
-        iter = chain(first_lines, data)
+        data.seek(0)
```

Comment:
Yes, but who guarantees that we should be processing the CSV from position 0?
When I made this change (seek -> chain) it wasn't the case. It was because we were using two csv readers - one to read the headers and then passing the stream to another csv reader to read the data. In that case, seeking to 0 would re-read the headers.

Therefore, I wouldn't seek to 0 - we can imagine a 'comment reader' that reads in the CSV until there are no more 'comment lines' (I remember seeing an issue about that once), or another who reads the 'BOM' character in the beginning of the stream. Then these processors pass the CSV reader the stream in the correct position to start, which is not 0.